pkg/cache/partition_info.go

```diff
@@ -715,55 +715,46 @@ func checkAndSetResource(resource *resources.Resource) string {
 	if resource != nil {
 		return strings.Trim(resource.String(), "map")
 	}
-	return ""
+	return "[]"
 }
 
-// TODO fix this:
-// should only return one element, only a root queue
-// remove hard coded values and unknown AbsUsedCapacity
-// map status to the draining flag
-func (pi *PartitionInfo) GetQueueInfos() []dao.QueueDAOInfo {
-	pi.RLock()
-	defer pi.RUnlock()
-
-	var queueInfos []dao.QueueDAOInfo
-
-	info := dao.QueueDAOInfo{}
-	info.QueueName = pi.Root.Name
-	info.Status = pi.Root.stateMachine.Current()
-	info.Capacities = dao.QueueCapacity{
-		Capacity:        checkAndSetResource(pi.Root.GetGuaranteedResource()),
-		MaxCapacity:     checkAndSetResource(pi.Root.GetMaxResource()),
-		UsedCapacity:    checkAndSetResource(pi.Root.GetAllocatedResource()),
-		AbsUsedCapacity: "20",
-	}
-	info.ChildQueues = GetChildQueueInfos(pi.Root)
-	queueInfos = append(queueInfos, info)
-
-	return queueInfos
-}
-
-// TODO fix this:
-// should only return one element, only a root queue
-// remove hard coded values and unknown AbsUsedCapacity
-// map status to the draining flag
-func GetChildQueueInfos(info *QueueInfo) []dao.QueueDAOInfo {
-	var infos []dao.QueueDAOInfo
-	for _, child := range info.children {
-		queue := dao.QueueDAOInfo{}
-		queue.QueueName = child.Name
-		queue.Status = child.stateMachine.Current()
-		queue.Capacities = dao.QueueCapacity{
-			Capacity:        checkAndSetResource(child.GetGuaranteedResource()),
-			MaxCapacity:     checkAndSetResource(child.GetMaxResource()),
-			UsedCapacity:    checkAndSetResource(child.GetAllocatedResource()),
-			AbsUsedCapacity: "20",
+func calculateAbsUsedCapacity(capacity *resources.Resource, usedCapacity *resources.Resource) *resources.Resource {
+	if capacity == nil {
+		return nil
+	}
+	if usedCapacity == nil {
+		return resources.NewResource()
+	}
+	absResource := make(map[string]resources.Quantity)
+	for resourceName, availableResource := range capacity.Resources {
+		var absResValue float32
+		if availableResource != 0 {
+			if usedResource, ok := usedCapacity.Resources[resourceName]; ok {
+				absResValue = float32(usedResource) / float32(availableResource) * 100
+			}
```

Comment:
Float divisions are 0 safe in go: it will return either +INF or -INF based on the signs.
We need to make sure that we handle all this over and underflow safe. If used = maxInt64, available is 10 then the abs value will overflow an Int64 on conversion...

If `availableResource` is 0 we set no usage. That seems counter intuitive to me.
The case that there is no capacity set but there is usage we skip it, again not logical.

We need to come up with a way to describe those cases:
- no capacity set, but usage is a 0
- capacity set to 0, usage is larger than 0

Reply:
- no capacity set, but usage is a 0: 
if no max capacity is set but we have resource usage I would return an empty resource, because we cannot calculate the absolute resource usage, since we don’t know what would mean 100%, so the web ui will show n/a

- capacity set to 0, usage is larger than 0: in my opinion this should not be a valid case. Since I am calculating the absolute usage according to the max capacity, what is a hard limit in my understanding. If we allow for the usage to exceed the max capacity, this limit will totally loose its sense. Are there cases when this is allowed?

Reply:
In the new changes et I used the following approach:
- in case of missing value for a resource type indifferently if it is the used or max capacity, I skipped the calculation
- capacity set to 0, usage is larger than 0: I made the calculation and returned maxInt
Please let me know if you have some better ideas for this cases.